README.md
# Collection Diff

Collection Diff allows you to compare 2 data collections to find operations to do: CREATE, UPDATE or DELETE data.

## Requirements

- PHP ^7.1

## Installation

```
composer install wizaplace/collection-diff
```

## Usage

If you want to compare data between database and form (like $_POST) 

```php
<?php

// Collection from your database
$from = [
    new FooObject("foo", "category1", 10.5),
    new FooObject("bar2", "category2", 9.5),
    new FooObject("bar1", "category1", 9.5),
    new FooObject("barFoo", "category1", 9.5),
    new FooObject("barFooBar", "category1", 9.5),
];

// Collection from you form like $_POST
$to = [
    [
        "name" => "foo",
        "category" => "category1",
        "price" => 11.5
    ],
    [
        "name" => "bar2",
        "category" => "category2",
        "price" => 9.5,
    ],
    [
        "name" => "bar1",
        "category" => "category2",
        "price" => 9.5,
    ],
    [
        "name" => "fooBar",
        "category" => "category3",
        "price" => 8.5,
    ],
];
```

You need to call `CollectionDiff::compare` to execute the comparison then call `CollectionDiff::getActions` to retrieve
actions.

`CollectionDiff::compare` first parameter is an array of primary keys needed by Collection Diff to execute the
comparison to check the concordance on this keys.

Your object will be normalized to be compared.

See example above.

### Symfony

If you want to use Collection Diff in your Symfony project, you can use the bridge 
`Wizaplace\CollectionDiff\Bridge\CollectionDiffBundle`.

```php
$container
    ->get(\Wizaplace\CollectionDiff\CollectionDiff::class)
    ->compare(['name', 'category'], $from, $to, false, true)
    ->getActions();
```

### Standalone

```php
$mySerializer = new FooSerializer();

$collectionDiff = new Wizaplace\CollectionDiff\CollectionDiff($mySerializer);
$collectionDiff->compare(['name', 'category'], $from, $to, false, true);

$collectionDiff->getActions();

// Or
$collectionDiff->getNothing();
$collectionDiff->getCreate();
$collectionDiff->getUpdate();
$collectionDiff->getDelete();
``` 

### Return the following results

```
array (size=4)
  1 => 
    array (size=1)
      0 => 
        array (size=3)
          'name' => string 'bar2' (length=4)
          'category' => string 'category2' (length=9)
          'price' => float 9.5
  2 => 
    array (size=1)
      0 => 
        array (size=3)
          'name' => string 'fooBar' (length=6)
          'category' => string 'category3' (length=9)
          'price' => float 8.5
  3 => 
    array (size=2)
      0 => 
        array (size=3)
          'name' => string 'foo' (length=3)
          'category' => string 'category1' (length=9)
          'price' => float 11.5
      1 => 
        array (size=3)
          'name' => string 'bar1' (length=4)
          'category' => string 'category2' (length=9)
          'price' => float 9.5
  4 => 
    array (size=1)
      0 => 
        array (size=3)
          'name' => string 'barFoo' (length=6)
          'category' => string 'category1' (length=9)
          'price' => float 9.5
```


CHANGELOG.md
# Changelog


## v1.0.0-beta.1

- Added `CollectionDiff::setNormalizers`
- Added CircleCI
- Updated `CollectionDiff::_construct` first param is now an array of `NormalizerInterface` 
- Updated `README.md`


## v1.0.0-beta.0

- First commit


tests/FooObject.php
<?php
/**
 * @author      Wizaplace DevTeam <dev@wizaplace.com>
 * @copyright   Copyright (c) Wizaplace
 * @license     MIT
 */

declare(strict_types=1);

namespace Wizaplace\Test\CollectionDiff;

class FooObject
{
    /** @var string */
    private $name;

    /** @var string */
    private $category;

    /** @var float */
    private $price;

    public function __construct(string $name, string $category, float $price)
    {
        $this->name = $name;
        $this->category = $category;
        $this->price = $price;
    }

    /**
     * @return string
     */
    public function getName(): string
    {
        return $this->name;
    }

    /**
     * @return string
     */
    public function getCategory(): string
    {
        return $this->category;
    }

    /**
     * @return float
     */
    public function getPrice(): float
    {
        return $this->price;
    }
}


tests/RandomObject.php
<?php
/**
 * @author      Wizaplace DevTeam <dev@wizaplace.com>
 * @copyright   Copyright (c) Wizaplace
 * @license     MIT
 */

declare(strict_types=1);

namespace Wizaplace\Test\CollectionDiff;

class RandomObject
{
    /** @var string */
    private $name;

    /** @var string */
    private $category;

    /** @var float */
    private $price;

    public function __construct(string $name, string $category, float $price)
    {
        $this->name = $name;
        $this->category = $category;
        $this->price = $price;
    }

    /**
     * @return string
     */
    public function getName(): string
    {
        return $this->name;
    }

    /**
     * @return string
     */
    public function getCategory(): string
    {
        return $this->category;
    }

    /**
     * @return float
     */
    public function getPrice(): float
    {
        return $this->price;
    }
}


tests/CollectionDiffTest.php
<?php
/**
 * @author      Wizaplace DevTeam <dev@wizaplace.com>
 * @copyright   Copyright (c) Wizaplace
 * @license     MIT
 */

declare(strict_types=1);

namespace Wizaplace\Test\CollectionDiff;

use PHPUnit\Framework\TestCase;
use Symfony\Component\Serializer\{
    Normalizer\PropertyNormalizer,
    Serializer
};
use Wizaplace\CollectionDiff\{
    CollectionDiff,
    Exception\CollectionDiffException
};

class CollectionDiffTest extends TestCase
{
    /** @var CollectionDiff */
    protected $collectionDiff;

    protected $from = [];
    protected $to = [];

    public function setUp(): void
    {
        $this->collectionDiff = new CollectionDiff(
            new Serializer([
                new PropertyNormalizer(),
            ])
        );

        $this->from = [
            new FooObject("foo", "category1", 10.5),
            new FooObject("bar2", "category2", 9.5),
            new FooObject("bar1", "category1", 9.5),
            new FooObject("barFoo", "category1", 9.5),
            new FooObject("barFooBar", "category1", 9.5),
        ];

        $this->to = [
            [
                "name" => "foo",
                "category" => "category1",
                "price" => 11.5
            ],
            [
                "name" => "bar2",
                "category" => "category2",
                "price" => 9.5,
            ],
            [
                "name" => "bar1",
                "category" => "category2",
                "price" => 9.5,
            ],
            [
                "name" => "fooBar",
                "category" => "category3",
                "price" => 8.5,
            ],
        ];
    }

    public function testCompareWithOldValues(): void
    {
        $this
            ->collectionDiff
            ->compare(['name', 'category'], $this->from, $this->to, true);

        static::assertCount(1, $this->collectionDiff->getNothing(), "[getNothing] should be equal to 1");
        static::assertSame($this->from[1], $this->collectionDiff->getNothing()[0]['newValues'], "[getNothing] is not equal to [from_1][newValues]");

        static::assertCount(1, $this->collectionDiff->getCreate(), "[getCreate] should be equal to 1");
        static::assertSame($this->to[3], $this->collectionDiff->getCreate()[0]['newValues'], "[getCreate] is not equal to [to_3][newValues]");
        static::assertSame($this->from[3], $this->collectionDiff->getCreate()[0]['oldValues'], "[getCreate] is not equal to [from_3][oldValues]");

        static::assertCount(2, $this->collectionDiff->getUpdate(), "[getUpdate] should be equal to 2");
        static::assertSame($this->to[0], $this->collectionDiff->getUpdate()[0]['newValues'], "[getUpdate] is not equal to [to_0][newValues]");
        static::assertSame($this->from[0], $this->collectionDiff->getUpdate()[0]['oldValues'], "[getUpdate] is not equal to [to_0][oldValues]");
        static::assertSame($this->to[2], $this->collectionDiff->getUpdate()[1]['newValues'], "[getUpdate] is not equal to [to_2][newValues]");
        static::assertSame($this->from[2], $this->collectionDiff->getUpdate()[1]['oldValues'], "[getUpdate] is not equal to [to_2][oldValues]");

        static::assertCount(2, $this->collectionDiff->getDelete(), "[getDelete] should be equal to 2");
        static::assertSame($this->from[3], $this->collectionDiff->getDelete()[0]['newValues'], "[getDelete] is not equal to [from_3][newValues]");
        static::assertSame($this->from[4], $this->collectionDiff->getDelete()[1]['newValues'], "[getDelete] is not equal to [from_4][newValues]");
    }

    public function testCompareWithoutOldValues(): void
    {
        $this
            ->collectionDiff
            ->compare(['name', 'category'], $this->from, $this->to);

        static::assertCount(1, $this->collectionDiff->getNothing(), "[getNothing] should be equal to 1");
        static::assertSame($this->from[1], $this->collectionDiff->getNothing()[0], "[getNothing] is not equal to [from_1]");

        static::assertCount(1, $this->collectionDiff->getCreate(), "[getCreate] should be equal to 1");
        static::assertSame($this->to[3], $this->collectionDiff->getCreate()[0], "[getCreate] is not equal to [to_3]");

        static::assertCount(2, $this->collectionDiff->getUpdate(), "[getUpdate] should be equal to 2");
        static::assertSame($this->to[0], $this->collectionDiff->getUpdate()[0], "[getUpdate] is not equal to [to_0]");
        static::assertSame($this->to[2], $this->collectionDiff->getUpdate()[1], "[getUpdate] is not equal to [to_2]");

        static::assertCount(2, $this->collectionDiff->getDelete(), "[getDelete] should be equal to 2");
        static::assertSame($this->from[3], $this->collectionDiff->getDelete()[0], "[getDelete] is not equal to [from_3]");
        static::assertSame($this->from[4], $this->collectionDiff->getDelete()[1], "[getDelete] is not equal to [from_4]");
    }

    public function testGetIncorrectMode(): void
    {
        static::expectException(CollectionDiffException::class);
        $this
            ->collectionDiff
            ->getActions(42);
    }
}


src/CollectionDiff.php
<?php
/**
 * @author      Wizaplace DevTeam <dev@wizaplace.com>
 * @copyright   Copyright (c) Wizaplace
 * @license     MIT
 */

declare(strict_types=1);

namespace Wizaplace\CollectionDiff;

use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
use Wizaplace\CollectionDiff\Exception\CollectionDiffException;

class CollectionDiff
{
    public const ACTION_NOTHING = 1;
    public const ACTION_CREATE = 2;
    public const ACTION_UPDATE = 3;
    public const ACTION_DELETE = 4;

    protected $actions = [
        self::ACTION_NOTHING => [],
        self::ACTION_CREATE  => [],
        self::ACTION_UPDATE  => [],
        self::ACTION_DELETE  => [],
    ];
    protected $primaryKeys = [];
    protected $from = [];
    protected $to = [];

    /** @var NormalizerInterface */
    protected $normalizer;

    public function __construct(NormalizerInterface $normalizer)
    {
        $this->normalizer = $normalizer;
    }

    public function compare(array $primaryKeys, array $from, array $to, bool $addOldValues = false, bool $strictComparison = true): self
    {
        $this->resetActions();

        $this->primaryKeys = $primaryKeys;

        $this->from = $from;
        $this->to = $to;

        $this->doCompare($addOldValues, $strictComparison);

        return $this;
    }

    public function getActions(int $action = null): array
    {
        if (false === is_null($action)) {
            if (false === array_key_exists($action, $this->actions)) {
                throw new CollectionDiffException("Unable to find this action.");
            }

            return $this->actions[$action];
        }

        return $this->actions;
    }

    public function getNothing(): array
    {
        return $this->getActions(static::ACTION_NOTHING);
    }

    public function getCreate(): array
    {
        return $this->getActions(static::ACTION_CREATE);
    }

    public function getUpdate(): array
    {
        return $this->getActions(static::ACTION_UPDATE);
    }

    public function getDelete(): array
    {
        return $this->getActions(static::ACTION_DELETE);
    }

    protected function doCompare(bool $addOldValues, bool $strictComparison): void
    {
        foreach ($this->to as $key => $values) {
            $from = $this->from[$key];

            if (true === array_key_exists($key, $this->from)) {
                /** @var array $normalizeValues */
                $normalizeValues = $this->normalizer->normalize($values);
                /** @var array $normalizeFrom */
                $normalizeFrom   = $this->normalizer->normalize($from);

                $action = static::ACTION_NOTHING;
                foreach ($this->primaryKeys as $primaryKey) {
                    if (
                        false === array_key_exists($primaryKey, $normalizeValues) ||
                        false === array_key_exists($primaryKey, $normalizeFrom)
                    ) {
                        continue;
                    }

                    if (
                        (
                            true === $strictComparison &&
                            $normalizeValues[$primaryKey] === $normalizeFrom[$primaryKey]
                        ) ||
                        (
                            false === $strictComparison &&
                            $normalizeValues[$primaryKey] == $normalizeFrom[$primaryKey]
                        )
                    ) {
                        foreach ($normalizeValues as $k => $v) {
                            if (false === array_key_exists($k, $normalizeFrom) || $v !== $normalizeFrom[$k]) {
                                $action = static::ACTION_UPDATE;

                                $this->add(static::ACTION_UPDATE, $values, $from, $addOldValues);
                                break;
                            }
                        }

                        if (static::ACTION_NOTHING !== $action) {
                            break;
                        }
                    } else {
                        $action = static::ACTION_DELETE;

                        $this->add(static::ACTION_DELETE, $from, null, $addOldValues);
                        $this->add(static::ACTION_CREATE, $values, $from, $addOldValues);

                        break;
                    }
                }

                if (static::ACTION_NOTHING === $action) {
                    $this->add(static::ACTION_NOTHING, $from, null, $addOldValues);
                }
            } else {
                $this->add(static::ACTION_CREATE, $values, null, $addOldValues);
            }

            unset($this->from[$key]);
        }

        if (count($this->from) > 0) {
            foreach ($this->from as $key => $from) {
                $this->add(static::ACTION_DELETE, $from, null, $addOldValues);
                unset($this->from[$key]);
            }
        }
    }

    /**
     * @param mixed      $newValues
     * @param mixed|null $oldValues
     */
    protected function add(int $action, $newValues, $oldValues = null, bool $addOldValues = false): void
    {
        if (true === $addOldValues) {
            $data = [
                'newValues' => $newValues,
            ];

            if (null !== $oldValues) {
                $data['oldValues'] = $oldValues;
            }
        } else {
            $data = $newValues;
        }

        $this->actions[$action][] = $data;
    }

    protected function resetActions(): void
    {
        $this->actions = [
            static::ACTION_NOTHING => [],
            static::ACTION_CREATE  => [],
            static::ACTION_UPDATE  => [],
            static::ACTION_DELETE  => [],
        ];
    }
}


demo/index.php
<?php
/**
 * @author      Wizaplace DevTeam <dev@wizaplace.com>
 * @copyright   Copyright (c) Wizaplace
 * @license     MIT
 */

declare(strict_types=1);

require __DIR__.'/../vendor/autoload.php';

use Symfony\Component\Serializer\Normalizer\PropertyNormalizer;
use Symfony\Component\Serializer\Serializer;
use Tracy\Debugger;
use Wizaplace\CollectionDiff\CollectionDiff;
use Wizaplace\Test\CollectionDiff\RandomObject;

// Enable Tracy
Debugger::enable(Debugger::DEVELOPMENT);
Debugger::$strictMode = true;
Debugger::$maxDepth = 15;
Debugger::$maxLength = 250;

// From your database
$from = [
    new RandomObject("foo", "category1", 10.5),
    new RandomObject("bar2", "category2", 9.5),
    new RandomObject("bar1", "category1", 9.5),
    new RandomObject("barFoo", "category1", 9.5),
    new RandomObject("barFooBar", "category1", 9.5),
];

// From you form like $_POST
$to = [
    [
        "name" => "foo",
        "category" => "category1",
        "price" => 11.5
    ],
    [
        "name" => "bar2",
        "category" => "category2",
        "price" => 9.5,
    ],
    [
        "name" => "bar1",
        "category" => "category2",
        "price" => 9.5,
    ],
    [
        "name" => "fooBar",
        "category" => "category3",
        "price" => 8.5,
    ],
];

$serializer = new Serializer([new PropertyNormalizer()]);

$collectionDiff = new CollectionDiff($serializer, ['name', 'category'], $from, $to);
dump($collectionDiff->getQueries());
